yjit/src/codegen.rs

```diff
@@ -2734,26 +2727,28 @@ fn gen_opt_aset(
         );
 
         // Call rb_ary_store
-        mov(cb, C_ARG_REGS[0], recv);
-        mov(cb, C_ARG_REGS[1], key);
-        sar(cb, C_ARG_REGS[1], uimm_opnd(1)); // FIX2LONG(key)
-        mov(cb, C_ARG_REGS[2], val);
+        let index = asm.load(key);
+        let index = asm.rshift(index, Opnd::UImm(1)); // FIX2LONG(key)
+        // Workaround to use rbx and avoid spill before moved by jit_prepare_routine_call
+        asm.mov(C_ARG_OPNDS[0], recv);
+        asm.mov(C_ARG_OPNDS[1], index);
+        asm.mov(C_ARG_OPNDS[2], val);
```

Comment:
We really shouldn't touch the C_ARG_OPNDS directly. Can you do the index computation after jit_prepare_routine_call instead?

Reply:
I got a little confused around this but now I figured it out https://github.com/Shopify/ruby/pull/387/commits/428f14a0f8787e26939b9f1b6a9c3c594551c125.